File: 백준/Gold/11401. 이항 계수 3/이항 계수 3.py
from sys import stdin
input = stdin.readline

def factorial(num):
    n = 1
    for i in range(2,num+1):
        n = (n * i) % mod
    return n

def mod_division(n,k):
    if k == 0:
        return 1
    elif k == 1:
        return n
    tmp = mod_division(n,k//2)
    if k % 2 == 1:
        return tmp * tmp * n % mod
    else:
        return tmp * tmp % mod

n,k = map(int, input().split())
mod = 1000000007
top = factorial(n)
bottom = factorial(n-k) * factorial(k) % mod
print(top * mod_division(bottom,mod-2) % mod)

File: 백준/Gold/11401. 이항 계수 3/이항 계수 3.java
import java.io.*;
import java.util.*;

public class Main {
	static long mod = 1000000007;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		long[] facto = new long[4000001];
		facto[0] = 1;
		for (int i = 1; i < 4000001; i++) {
			facto[i] = facto[i - 1] * i % mod;
		}
		st = new StringTokenizer(br.readLine());
		int N = Integer.parseInt(st.nextToken());
		int R = Integer.parseInt(st.nextToken());
		long top = facto[N];
		long bottom = facto[N - R] * facto[R] % mod;
		System.out.println(top * power(bottom, mod - 2) % mod);
	}

	static long power(long n, long k) { // N^K
		long res = 1L;
		n %= mod;
		while (k > 0) {
			if (k % 2 == 1) {
				res *= n;
				res %= mod;
			}
			k >>= 1;
			n *= n;
			n %= mod;
		}
		return res;
	}
}